src/bpftrace.cpp

```diff
@@ -104,237 +107,111 @@ Probe BPFtrace::generate_probe(const ast::AttachPoint &ap, const ast::Probe &p)
-          !p.need_expansion && attach_funcs.size()) {
-        if (!has_wildcard(attach_point->target)) {
-          auto probe = generate_probe(*attach_point, p);
-          probe.funcs = attach_funcs;
-          resources.probes.push_back(probe);
-          continue;
+    if (has_wildcard(ap.target)) {
+      // If we have a wildcard in the target path, we need to generate one
+      // probe per expanded target.
+      assert(type == ProbeType::uprobe || type == ProbeType::uretprobe);
+      std::unordered_map<std::string, Probe> target_map;
+      for (const auto &func : matches) {
+        ast::AttachPoint match_ap = ap.create_expansion_copy(func);
+        // Use the original (possibly wildcarded) function name
+        match_ap.func = ap.func;
+        auto found = target_map.find(match_ap.target);
+        if (found != target_map.end()) {
+          found->second.funcs.push_back(func);
         } else {
-          // If we have a wildcard in the target path, we need to generate one
-          // probe per expanded target.
-          std::unordered_map<std::string, Probe> target_map;
-          for (const auto &func : attach_funcs) {
-            ast::AttachPoint ap = attach_point->create_expansion_copy(func);
-            // Use the original (possibly wildcarded) function name
-            ap.func = attach_point->func;
-            auto found = target_map.find(ap.target);
-            if (found != target_map.end()) {
-              found->second.funcs.push_back(func);
-            } else {
-              auto probe = generate_probe(ap, p);
-              probe.funcs.push_back(func);
-              target_map.insert({ { ap.target, probe } });
-            }
-          }
-          for (auto &pair : target_map) {
-            resources.probes.push_back(std::move(pair.second));
-          }
-          continue;
+          auto probe = generate_probe(match_ap, p);
+          probe.funcs.push_back(func);
+          target_map.insert({ { match_ap.target, probe } });
         }
       }
-    } else if ((probetype(attach_point->provider) == ProbeType::uprobe ||
-                probetype(attach_point->provider) == ProbeType::uretprobe ||
-                probetype(attach_point->provider) == ProbeType::watchpoint ||
-                probetype(attach_point->provider) ==
-                    ProbeType::asyncwatchpoint) &&
-               !attach_point->func.empty()) {
-      std::set<std::string> matches;
-
-      struct symbol sym = {};
-      int err = resolve_uname(attach_point->func, &sym, attach_point->target);
-
-      if (attach_point->lang == "cpp") {
-        // As the C++ language supports function overload, a given function name
-        // (without parameters) could have multiple matches even when no
-        // wildcards are used.
-        matches = probe_matcher_->get_matches_for_ap(*attach_point);
+      for (auto &pair : target_map) {
+        resources.probes.push_back(std::move(pair.second));
       }
```

Comment:
Is this ever executed? It looks like target expansion takes place in `SemanticAnalyser`, so `has_wildcard(ap.target)` is always false.

Reply:
Semantic analyser only resolves the binary path. The comment there says:
https://github.com/bpftrace/bpftrace/blob/ce8f81a1945ac23ed94fe0db942c9592971a6320/src/ast/passes/semantic_analyser.cpp#L2605-L2607
That is, when wildcards are used (and multiple matches exist), the wildcard is not expanded.

I double checked and this is indeed executed when attaching to `uprobe:tests/testprogs/uprobe*:*`.